File: edge/serving-worker/src/manifest.ts
// SPDX-License-Identifier: FSL-1.1-Apache-2.0
//
// Deploy-manifest model + path resolution — the pure logic that turns a request
// path into a content-addressed blob key. Side-effect-light and unit-testable
// without a live edge (KV/R2 are injected as minimal interfaces in index.ts).
//
// Object layout (content-addressed, dedup — docs/ARCHITECTURE.md §3):
//   - per-deploy manifest JSON at  manifests/<org_id>/<site_id>/<version_id>.json
//     mapping each served request-path → { sha256, content_type }.
//   - blobs at                      blobs/<org_id>/<sha256>
// Serving resolves path → sha256 via the manifest, then streams the blob. A
// version is one immutable manifest; the blobs it references are shared across
// deploys (and orgs are isolated by the per-org prefix).

import type { RouteValue } from "./route";

/**
 * One manifest entry: the content-addressed identity of a served file plus the
 * Content-Type the Go API recorded at publish time (authoritative — we do NOT
 * re-sniff untrusted tenant bytes at the edge).
 */
export interface ManifestEntry {
  /** Lowercase hex SHA-256 of the blob (also its R2 key suffix). */
  sha256: string;
  /** Content-Type recorded at publish; served verbatim. */
  content_type: string;
  /** Decoded byte length, when the Go API records it (optional). */
  size?: number;
}

/**
 * A published deploy manifest. `files` maps a normalized, prefix-relative
 * request path (no leading slash) to its entry — e.g. "index.html",
 * "assets/app.4f3a9c2b.js", "blog/index.html". Carries `schema_version` so the
 * Worker can refuse a shape it does not understand (mirrors the KV contract).
 */
export interface Manifest {
  /** Contract version of this manifest shape; the Worker pins what it accepts. */
  schema_version: number;
  /** Normalized request path → content-addressed entry. */
  files: Record<string, ManifestEntry>;
}

/** The manifest-shape version this Worker understands. */
export const SUPPORTED_MANIFEST_SCHEMA_VERSION = 1 as const;

/** Lowercase-hex SHA-256, exactly 64 chars (R2 blob-key safety). */
const SHA256_RE = /^[0-9a-f]{64}$/;

/**
 * The R2 key of a version's deploy manifest. One immutable JSON per published
 * version, namespaced org → site → version.
 *
 *   manifests/<org_id>/<site_id>/<version_id>.json
 */
export function manifestKey(route: RouteValue): string {
  return `manifests/${route.org_id}/${route.site_id}/${route.version_id}.json`;
}

/**
 * The R2 key of a content-addressed blob for the route's org. Blobs are shared
 * (dedup) within an org; the org prefix is the isolation boundary.
 *
 *   blobs/<org_id>/<sha256>
 */
export function blobKey(orgId: string, sha256: string): string {
  return `blobs/${orgId}/${sha256}`;
}

/**
 * Validate an untrusted manifest JSON into a typed `Manifest`. Returns null on
 * any shape/version mismatch (or a malformed entry) so callers fail closed
 * (404) rather than serving from a corrupt projection. The Worker pins
 * `schema_version` exactly.
 */
export function parseManifest(raw: unknown): Manifest | null {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return null;
  const v = raw as Record<string, unknown>;

  if (v.schema_version !== SUPPORTED_MANIFEST_SCHEMA_VERSION) return null;
  if (v.files === null || typeof v.files !== "object" || Array.isArray(v.files)) {
    return null;
  }

  const files: Record<string, ManifestEntry> = {};
  for (const [path, rawEntry] of Object.entries(v.files as Record<string, unknown>)) {
    const entry = parseEntry(rawEntry);
    if (entry === null) return null; // any bad entry → reject the whole manifest
    files[path] = entry;
  }

  return { schema_version: v.schema_version, files };
}

function parseEntry(raw: unknown): ManifestEntry | null {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return null;
  const e = raw as Record<string, unknown>;

  if (typeof e.sha256 !== "string" || !SHA256_RE.test(e.sha256)) return null;
  if (typeof e.content_type !== "string" || e.content_type.length === 0) return null;

  const entry: ManifestEntry = { sha256: e.sha256, content_type: e.content_type };
  if (typeof e.size === "number" && Number.isFinite(e.size) && e.size >= 0) {
    entry.size = e.size;
  }
  return entry;
}

/**
 * Resolve a (already cleaned, prefix-relative) request path to its manifest
 * entry, applying static-site fallbacks IN ORDER:
 *
 *  - Directory request (empty path, or trailing "/") → "<path>index.html".
 *  - Exact match → that entry.
 *  - Extension-less "pretty" path "/about" → "about/index.html", then "about.html".
 *
 * The matched key is returned alongside its entry so the caller can apply the
 * Cache-Control policy by the served path's extension (HTML short, hashed
 * assets immutable). Returns null when nothing in the manifest matches.
 */
export function resolveManifestEntry(
  manifest: Manifest,
  cleanRelPath: string,
): { path: string; entry: ManifestEntry } | null {
  for (const candidate of candidatePaths(cleanRelPath)) {
    const entry = manifest.files[candidate];
    if (entry !== undefined) return { path: candidate, entry };
  }
  return null;
}

/**
 * Ordered manifest-path candidates for a cleaned request path. Mirrors the
 * directory-index + pretty-URL semantics of the old key resolution, but now
 * resolved against the manifest's path set rather than probed against R2.
 */
export function candidatePaths(cleanRelPath: string): string[] {
  // Directory request (root or trailing slash) → index.html only.
  if (cleanRelPath === "" || cleanRelPath.endsWith("/")) {
    return [`${cleanRelPath}index.html`];
  }

  const candidates: string[] = [cleanRelPath];

  // Extension-less "pretty" path → also try directory index and `.html`.
  if (!hasExtension(cleanRelPath)) {
    candidates.push(`${cleanRelPath}/index.html`);
    candidates.push(`${cleanRelPath}.html`);
  }

  return candidates;
}

/** The manifest path for a version's custom 404 page, if it ships one. */
export const NOT_FOUND_PATH = "404.html";

/** True if the final path segment carries a file extension (e.g. `.css`). */
function hasExtension(cleanRelPath: string): boolean {
  const last = cleanRelPath.split("/").pop() ?? "";
  const dot = last.lastIndexOf(".");
  return dot > 0 && dot < last.length - 1;
}

File: services/serve/internal/manifest/manifest.go
// SPDX-License-Identifier: FSL-1.1-Apache-2.0
//
// Package manifest is the Go port of the serving Worker's per-deploy manifest
// model + path resolution (edge/serving-worker/src/manifest.ts). It turns a
// cleaned request path into a content-addressed blob key via the published
// manifest, with directory-index + pretty-URL fallbacks, and fails CLOSED on any
// drift (schema_version != 1, bad sha256, empty content_type).
package manifest

import (
	"bytes"
	"encoding/json"
	"regexp"
	"strings"
)

// SupportedSchemaVersion is the manifest-shape version this server understands.
// It MUST equal the Worker's SUPPORTED_MANIFEST_SCHEMA_VERSION and
// internal/manifest.SchemaVersion (== 1).
const SupportedSchemaVersion = 1

// NotFoundPath is the manifest key of a version's custom 404 page, if it ships one.
const NotFoundPath = "404.html"

var sha256RE = regexp.MustCompile(`^[0-9a-f]{64}$`)

// Entry is one validated manifest entry: the content-addressed sha256, the
// authoritative content_type recorded at publish, and an optional decoded size.
type Entry struct {
	SHA256      string
	ContentType string
	Size        int64
	HasSize     bool
}

// Manifest is a parsed, validated deploy manifest: request-path → entry.
type Manifest struct {
	Files map[string]Entry
}

// rawManifest / rawEntry mirror the stored JSON shape (deployments.go
// storedManifest). schema_version + size use json.Number so we can enforce the
// "integer" and ">=0 finite" constraints faithfully.
type rawManifest struct {
	SchemaVersion *json.Number               `json:"schema_version"`
	Files         map[string]json.RawMessage `json:"files"`
}

type rawEntry struct {
	SHA256      *string      `json:"sha256"`
	ContentType *string      `json:"content_type"`
	Size        *json.Number `json:"size"`
}

// Parse validates untrusted manifest JSON into a Manifest, returning ok=false on
// any shape/version mismatch or a single malformed entry (the WHOLE manifest is
// rejected, mirroring parseManifest). Callers fail closed (404) on ok=false.
//
// NOTE on content_type: the Go publisher writes content_type with `omitempty`, so
// an entry CAN omit it; we follow the WORKER and REJECT a missing/empty
// content_type (fail closed) to stay byte-for-byte faithful.
func Parse(raw []byte) (Manifest, bool) {
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	var rm rawManifest
	if err := dec.Decode(&rm); err != nil {
		return Manifest{}, false
	}
	if rm.SchemaVersion == nil {
		return Manifest{}, false
	}
	sv, err := rm.SchemaVersion.Int64()
	if err != nil || sv != SupportedSchemaVersion {
		return Manifest{}, false
	}
	if rm.Files == nil {
		return Manifest{}, false
	}

	files := make(map[string]Entry, len(rm.Files))
	for path, rawEntry := range rm.Files {
		entry, ok := parseEntry(rawEntry)
		if !ok {
			return Manifest{}, false // any bad entry → reject the whole manifest
		}
		files[path] = entry
	}
	return Manifest{Files: files}, true
}

func parseEntry(raw json.RawMessage) (Entry, bool) {
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	var re rawEntry
	if err := dec.Decode(&re); err != nil {
		return Entry{}, false
	}
	if re.SHA256 == nil || !sha256RE.MatchString(*re.SHA256) {
		return Entry{}, false
	}
	if re.ContentType == nil || *re.ContentType == "" {
		return Entry{}, false
	}
	entry := Entry{SHA256: *re.SHA256, ContentType: *re.ContentType}
	if re.Size != nil {
		n, err := re.Size.Int64()
		if err == nil && n >= 0 {
			entry.Size = n
			entry.HasSize = true
		}
	}
	return entry, true
}

// Match is a resolved manifest entry plus the manifest key it matched (the served
// path, which drives Cache-Control).
type Match struct {
	Path  string
	Entry Entry
}

// Resolve resolves a cleaned, prefix-relative request path to a manifest entry,
// applying the static-site fallbacks in order. Returns ok=false when nothing
// matches. Faithful port of resolveManifestEntry/candidatePaths.
func (m Manifest) Resolve(cleanRelPath string) (Match, bool) {
	for _, candidate := range candidatePaths(cleanRelPath) {
		if entry, ok := m.Files[candidate]; ok {
			return Match{Path: candidate, Entry: entry}, true
		}
	}
	return Match{}, false
}

// NotFoundEntry returns the version's custom 404.html entry, if present.
func (m Manifest) NotFoundEntry() (Entry, bool) {
	e, ok := m.Files[NotFoundPath]
	return e, ok
}

// candidatePaths is a faithful port of candidatePaths: directory request → only
// "<path>index.html"; exact match; extension-less pretty path also tries
// "<path>/index.html" then "<path>.html".
func candidatePaths(cleanRelPath string) []string {
	if cleanRelPath == "" || strings.HasSuffix(cleanRelPath, "/") {
		return []string{cleanRelPath + "index.html"}
	}
	candidates := []string{cleanRelPath}
	if !hasExtension(cleanRelPath) {
		candidates = append(candidates, cleanRelPath+"/index.html", cleanRelPath+".html")
	}
	return candidates
}

// hasExtension reports whether the final path segment carries a file extension
// (a '.' that is neither the first nor the last char). Mirrors hasExtension.
func hasExtension(cleanRelPath string) bool {
	last := lastSegment(cleanRelPath)
	dot := strings.LastIndex(last, ".")
	return dot > 0 && dot < len(last)-1
}

func lastSegment(p string) string {
	if i := strings.LastIndex(p, "/"); i != -1 {
		return p[i+1:]
	}
	return p
}
